In [ ]:
# install the library
!pip install transformers

     |████████████████████████████████| 2.1MB 20.5MB/s 
     |████████████████████████████████| 901kB 52.9MB/s 
     |████████████████████████████████| 3.3MB 52.0MB/s 


In [ ]:
# import packages
import torch
from transformers import AutoModel, AutoTokenizer, BertTokenizer

set gradient calculation off in context-manager.[more about gradient decsent from pytorch perspective](https://pytorch.org/tutorials/beginner/blitz/autograd_tutorial.html)

In [ ]:
torch.set_grad_enabled(False)

In [ ]:
# apply a pretrained bert model
# initialize the model

model = AutoModel.from_pretrained('bert-base-cased')

# set up tokenizers that will let us map text input to a sequence of integers, so that the model can manipulate with them. 
# initialize the tokenizer

tokenizer = AutoTokenizer.from_pretrained('bert-base-cased')

In [ ]:
# ask the tokenizer to return a pytorch type tensor

tokens_pt = tokenizer('text input example here', return_tensors='pt')

In [ ]:
# inspect the variable

tokens_pt

{'input_ids': tensor([[ 101, 3087, 7758, 1859, 1303,  102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1]])}

input_ids: tokenized text inputs, also beginning and end of the string.

token_type_ids: map every token to their corresponding segment.

attention_mask: used for masking mechanism.



In [ ]:
# feed the tensor to the model created

output = model(**tokens_pt)

In [ ]:
# inspect the output tensor 'last_hidden_state'

output.last_hidden_state

tensor([[[ 0.2156, -0.0028,  0.3592,  ...,  0.1299,  0.4936,  0.5782],
         [ 0.0879, -0.0520,  0.4663,  ...,  0.0022,  0.3878,  0.7273],
         [ 0.2973,  0.1926,  0.1119,  ...,  0.2808,  0.0537,  0.7051],
         [ 0.2185,  0.0711,  0.1865,  ...,  0.3129, -0.0850,  0.5330],
         [ 0.4345,  0.0094,  0.3702,  ...,  0.2957,  0.8259,  0.2300],
         [ 0.3789,  0.7589, -0.0261,  ..., -0.0082,  0.9710,  0.4226]]])

In [ ]:
output.last_hidden_state.shape

torch.Size([1, 6, 768])

the above tensor shape tells us that for every token in the input, the model has generated a representation (768 elements in 1 list, and there are 6 lists in total). this token based representation is useful for tasks that require token-level operation, such as question answering.

In [ ]:
output.pooler_output

tensor([[-0.7490,  0.5348,  0.9999, -0.9950,  0.9683,  0.7019,  0.9819, -0.9204,
         -0.9736, -0.7417,  0.9847,  0.9987, -0.9906, -0.9999,  0.3985, -0.9665,
          0.9907, -0.6191, -1.0000, -0.0430, -0.4853, -0.9999,  0.2182,  0.9104,
          0.9724,  0.0028,  0.9898,  1.0000,  0.8888,  0.3369,  0.3485, -0.9940,
          0.5771, -0.9995,  0.2940, -0.0457,  0.4327, -0.3162,  0.7336, -0.9058,
         -0.8086,  0.0375, -0.0352, -0.5316,  0.6694,  0.4234,  0.2938, -0.0670,
         -0.1897,  1.0000, -0.9688,  1.0000, -0.9741,  0.9992,  0.9960,  0.5325,
          0.9963,  0.2593, -0.9890,  0.4152,  0.9753,  0.1195,  0.9297, -0.1981,
         -0.3970, -0.5205, -0.7460,  0.2381, -0.6901,  0.6053,  0.2539,  0.3921,
          0.9821, -0.9466, -0.0156, -0.9524,  0.6298, -1.0000,  0.9620,  1.0000,
          0.0437, -0.9997,  0.9951, -0.3418, -0.0544, -0.2435, -0.9942, -0.9996,
          0.0773, -0.5973,  0.7423, -0.9901,  0.4142, -0.6553,  1.0000, -0.7284,
         -0.3468,  0.4330,  

In [ ]:
output.pooler_output.shape

torch.Size([1, 768])

the above tensor shape tells us that for the whole input, the model has generated a representation (768 elements in 1 list). this is useful if we need to extract the overall context of the string(sequence) but not on word level. such as sentiment analysis, or information retrieval.


In [ ]:
from transformers import TFBertModel, BertModel

model_pt = BertModel.from_pretrained('bert-base-cased')

from transformers import DistilBertModel

bert_distil = DistilBertModel.from_pretrained('distilbert-base-cased')
input_pt = tokenizer(
    'This is a sample input to demonstrate performance of distiled models especially inference time', 
    return_tensors="pt"
)


%time _ = bert_distil(input_pt['input_ids'])
%time _ = model_pt(input_pt['input_ids'])

CPU times: user 44.1 ms, sys: 0 ns, total: 44.1 ms
Wall time: 44.1 ms
CPU times: user 87.2 ms, sys: 0 ns, total: 87.2 ms
Wall time: 86.9 ms
